File: 백준/Gold/14502. 연구소/연구소.cc
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

int n, m;
vector<vector<int> > lab;

void bfs(vector<vector<int> > &visited) {
    queue<pair<int, int> > q;
    vector<int> dx = {1, -1, 0, 0};
    vector<int> dy = {0, 0, 1, -1};
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (lab[i][j] == 2) {
                q.push({i, j});
                visited[i][j] = true;
            }
        }
    }
    while (!q.empty()) {
        auto [x, y] = q.front(); q.pop();
        for (int d = 0; d < 4; d++) {
            int nx = x + dx[d];
            int ny = y + dy[d];
            if (nx >= 0 && ny >= 0 && nx < n && ny < m && !visited[nx][ny] && lab[nx][ny] == 0) {
                visited[nx][ny] = true;
                q.push({nx, ny});
            }
        }
    }
}

int get_safe_area() {
    vector<vector<int>> visited(n, vector<int>(m, false));
    bfs(visited);
    int safe_area = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (lab[i][j] == 0 && !visited[i][j]) {
                safe_area++;
            }
        }
    }
    return safe_area;
}

int main() {
    cin >> n >> m;
    lab.resize(n, vector<int>(m));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            cin >> lab[i][j];
        }
    }
    int max_area = 0;
    vector<pair<int, int>> empty_spaces;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (lab[i][j] == 0) {
                empty_spaces.push_back({i, j});
            }
        }
    }
    int size = empty_spaces.size();
    for (int i = 0; i < size; i++) {
        for (int j = i + 1; j < size; j++) {
            for (int k = j + 1; k < size; k++) {
                auto [x1, y1] = empty_spaces[i];
                auto [x2, y2] = empty_spaces[j];
                auto [x3, y3] = empty_spaces[k];
                lab[x1][y1] = lab[x2][y2] = lab[x3][y3] = 1;
                max_area = max(max_area, get_safe_area());
                lab[x1][y1] = lab[x2][y2] = lab[x3][y3] = 0;
            }
        }
    }
    cout << max_area << endl;
    return 0;
}

File: 백준/Gold/14502. 연구소/연구소.py
from itertools import combinations

def bfs(lab, visited, n, m):
    directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
    queue = []
    for i in range(n):
        for j in range(m):
            if lab[i][j] == 2:
                queue.append((i, j))
                visited[i][j] = True
    while queue:
        x, y = queue.pop(0)
        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if 0 <= nx < n and 0 <= ny < m and not visited[nx][ny] and lab[nx][ny] == 0:
                visited[nx][ny] = True
                queue.append((nx, ny))

def get_safe_area(lab, n, m):
    visited = [[False] * m for _ in range(n)]
    bfs(lab, visited, n, m)
    return sum(1 for i in range(n) for j in range(m) if lab[i][j] == 0 and not visited[i][j])

def max_safe_area(n, m, lab):
    empty_spaces = [(i, j) for i in range(n) for j in range(m) if lab[i][j] == 0]
    max_area = 0
    for walls in combinations(empty_spaces, 3):
        for x, y in walls:
            lab[x][y] = 1
        max_area = max(max_area, get_safe_area(lab, n, m))
        for x, y in walls:
            lab[x][y] = 0
    return max_area

n, m = map(int, input().split())
lab = [list(map(int, input().split())) for _ in range(n)]
print(max_safe_area(n, m, lab))